### DISCUSSION TITLE
How to add an NXP HAL module to my project?

### DISCUSSION THREAD

**Discussion by notronrj** [zephyrproject-rtos/zephyr#71571] [Q&A] [answered]
I'm building an application for an NXP device that needs to use the  `./modules/hal/nxp/mcux/mcux-sdk/drivers/ocotp/` driver to read and write the OTP e-fuses.

What magical CMake incantations do I need to add to my application level CMakeLists.txt file to include that module?

I've looked at several example CMakeLists.txt files etc and googled but I'm just not experienced with CMake enough to make sense of it all.

Thank you.

**Comment by DerekSnell** [comment, marked as answer]
Edited: This discussion is no longer up to date with the latest HAL_NXP.  These details still work to include a driver file, and are still useful for versions of Zephyr v4.1 or older, or when working with a device that is not supported in the latest MCUXpresso SDK.  For the latest details, see [Using MCUXpresso SDK drivers in a Zephyr application](https://community.nxp.com/t5/Zephyr-Project-Knowledge-Base/Using-MCUXpresso-SDK-drivers-in-a-Zephyr-application/ta-p/2077778).

Many drivers in the HAL are enabled for Zephyr and included with Kconfigs.  The enabled drivers are included from the [hal_nxp.cmake](https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/hal_nxp.cmake) file.  But the HAL includes many other [SDK drivers](https://github.com/zephyrproject-rtos/hal_nxp/tree/master/mcux/mcux-sdk/drivers) that are not yet enabled in Zephyr.  These drivers can be included by the application, and the SDK driver APIs can be called directly from the application.

This example uses the [PUF](https://github.com/zephyrproject-rtos/hal_nxp/tree/master/mcux/mcux-sdk/drivers/puf) driver, a security feature on many NXP SOCs, including the LPC55S69.  To include this driver, add the following lines to the Zephyr application's CmakeLists.txt file after the command `find_package(Zephyr ...)`:
```
# Include the fsl_puf driver in HAL_NXP
set(PUF_path $ENV{ZEPHYR_BASE}/../modules/hal/nxp/mcux/mcux-sdk/drivers/puf)
zephyr_sources(${PUF_path}/fsl_puf.c)
zephyr_include_directories(${PUF_path})
```

**Comment by HyperionCoding** [comment]
Continuing discussion here as it seems fitting.

I'm using Zephyr 3.7 where the above mentioned way indeed works, except in one case. `driver_prince` has other includes inside `driver_prince.cmake` which cmake cannot find

```
include could not find requested file:
    driver_puf
...
include could not find requested file:
    driver_iap1
```

That can be solved by including PUF and IAP1 drivers in the same way as above. However, the next break is a missing variable

```
fsl_prince.h:13:10: error: #include expects "FILENAME" or <FILENAME>
   13 | #include FFR_INCLUDE
```

This is defined in `set_device_LPC55S69.cmake` which is never actually used as far as I can tell.

So the question: is there a better way to include modules without Zephyr support? v4.2 does change it but I'd like to hear if anyone knows how to solve this "properly".
